[FakeTensor] Reuse flat_args throughout FakeTensorMode.dispatch

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112493
* __->__ #112418

This function repeatedly flattens and unflattens the `args, kwargs` pair so we
get a quite significant perf improvement from saving the `flat_args` and
operating directly on those. I see a 15% improvement in dispatch for
`empty_strided`.

peterbell10 committed:
Reuse flat_args throughout FakeTensorMode.dispatch

This function repeatedly flattens and unflattens the `args, kwargs` pair so we
get a quite significant perf improvement from saving the `flat_args` and
operating directly on those. I see a 15% improvement in dispatch for
`empty_strided`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Reuse flat_args throughout FakeTensorMode.dispatch"


This function repeatedly flattens and unflattens the `args, kwargs` pair so we
get a quite significant perf improvement from saving the `flat_args` and
operating directly on those. I see a 15% improvement in dispatch for
`empty_strided`.

[ghstack-poisoned]

lezcano commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1732,23 +1739,19 @@ def validate(x):
 
-    def wrap_meta_outputs_with_default_device_logic(self, r, func, args, kwargs):
-        wrap = self.gen_wrap_fn(func, args, kwargs)
+    def wrap_meta_outputs_with_default_device_logic(self, r, func, flat_args, kwargs):
```
nit. perhaps just pass `device rather than passing flat_args *and* kwargs. I found this a bit confusing.

lezcano commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1927,7 +1933,8 @@ def map_out(e):
 
-    return tree_map(map_out, r)
+    flat_out = [map_out(o) for o in flat_out]
+    return pytree.tree_unflatten(flat_out, out_spec)
```
Here perhaps we just want to `tree_map` all the transformation on `r` rather than unpacking and packing again?

lezcano reviewed (approved):
Two optional nits

peterbell10 committed:
Update on "[FakeTensor] Reuse flat_args throughout FakeTensorMode.dispatch"


This function repeatedly flattens and unflattens the `args, kwargs` pair so we
get a quite significant perf improvement from saving the `flat_args` and
operating directly on those. I see a 15% improvement in dispatch for
`empty_strided`.

[ghstack-poisoned]

peterbell10 committed:
Update on "[FakeTensor] Reuse flat_args throughout FakeTensorMode.dispatch"


This function repeatedly flattens and unflattens the `args, kwargs` pair so we
get a quite significant perf improvement from saving the `flat_args` and
operating directly on those. I see a 15% improvement in dispatch for
`empty_strided`.

[ghstack-poisoned]
